### PR TITLE
176372 retry mechanism for endpoints

### PR DESCRIPTION
**Pull request by mayur1501364**
Consist changes of [PBI 176372](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/176372)

- Added retry mechanism
- Added unit test cases

### CODE REVIEW

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/UserPermitServiceTests.cs [hunk 1]
```diff
@@ -156,5 +164,42 @@ public void WhenUserPermitServiceResponseOtherThanOk_ThenResponseShouldNotBeOk(H
                 && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Request to UserPermitService GET {RequestUri} started"
             ).MustHaveHappenedOnceExactly();
         }
+
+        [Test]
+        [TestCase(HttpStatusCode.TooManyRequests, "TooManyRequests")]
+        public void WhenUserPermitServiceResponseTooManyRequest_ThenResponseShouldNotBeOk(HttpStatusCode statusCode, string content)
```
Comment (shirinbt14926) [inline tests/UKHO.S100PermitService.Common.UnitTests/Services/UserPermitServiceTests.cs:170]:
```suggestion
        public void WhenUserPermitServiceResponseTooManyRequests_ThenResponseShouldNotBeOk(HttpStatusCode statusCode, string content)
```

**Review by shirinbt14926 — COMMENTED** [review]
Minor comments please

File: src/UKHO.S100PermitService.Common/Handlers/WaitAndRetryPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+            return Policy.HandleResult<HttpResponseMessage>(res => res.StatusCode == HttpStatusCode.ServiceUnavailable ||
+                             res.StatusCode == HttpStatusCode.TooManyRequests ||
+                             res.StatusCode == HttpStatusCode.InternalServerError).WaitAndRetry(retryCount, _ => TimeSpan.FromSeconds(sleepDuration),
+                             onRetry: (response, timespan, retryAttempt, context) =>
+                             {
+                                 var correlationId = response.Result.RequestMessage!.Headers.FirstOrDefault(h => h.Key.ToLowerInvariant() == "x-correlation-id");
+                                 var retryAfter = 0;
+                                 logger
+                                 .LogInformation(eventId.ToEventId(), "Re-trying service request with uri {RequestUri} and delay {delay}ms and retry attempt {retry} with _X-Correlation-ID:{correlationId} as previous request was responded with {StatusCode}.",
```
[the remark this answers is not in the document]
Comment (mayur1501364) [reply src/UKHO.S100PermitService.Common/Handlers/WaitAndRetryPolicy.cs:33]: Changed the log

File: src/UKHO.S100PermitService.Common/Events/EventIds.cs [hunk 3]
```diff
@@ -153,6 +153,21 @@ public enum EventIds
...
+
+        /// <summary>
+        /// 840032 - Request for retrying user permit api endpoint
+        /// </summary>
+        RetryHttpClientUserPermitRequest = 840032,
+
+        /// <summary>
+        /// 840033 - Request for retrying product key service api endpoint
+        /// </summary>
+        RetryHttpClientPKSRequest = 840033
```
[the remark this answers is not in the document]
Comment (mayur1501364) [reply src/UKHO.S100PermitService.Common/Events/EventIds.cs:171]: Changed event name

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/UserPermitServiceTests.cs (hunk 1 above)
Comment (mayur1501364) [reply tests/UKHO.S100PermitService.Common.UnitTests/Services/UserPermitServiceTests.cs:170]: changed test case name

**Review by shirinbt14926 — APPROVED** [review]
LGTM. Thanks @mayur1501364  for making changes

**Review by harshal11869 — APPROVED** [review]
Looks Good,

File: src/UKHO.S100PermitService.Common/Handlers/WaitAndRetryPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        public RetryPolicy<HttpResponseMessage> GetRetryPolicy(ILogger logger, EventIds eventId)
+        {
+            var retryCount = int.Parse(_waitAndRetryConfiguration.Value.RetryCount);
+            var sleepDuration = double.Parse(_waitAndRetryConfiguration.Value.SleepDurationInSeconds);
+
+            return Policy.HandleResult<HttpResponseMessage>(res => res.StatusCode == HttpStatusCode.ServiceUnavailable ||
+                             res.StatusCode == HttpStatusCode.TooManyRequests).WaitAndRetry(retryCount, _ => TimeSpan.FromSeconds(sleepDuration),
+                             onRetry: (response, timespan, retryAttempt, context) =>
+                             {
+                                 var correlationId = response.Result.RequestMessage!.Headers.FirstOrDefault(h => h.Key.ToLowerInvariant() == "x-correlation-id");
```
Comment (harshal11869) [inline src/UKHO.S100PermitService.Common/Handlers/WaitAndRetryPolicy.cs:29]:
```suggestion
                                 var correlationId = response.Result.RequestMessage!.Headers.FirstOrDefault(h => h.Key.ToLowerInvariant() == PermitServiceConstants.XCorrelationIdHeaderKey.ToLowerInvariant());
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: src/UKHO.S100PermitService.Common/Handlers/WaitAndRetryPolicy.cs (hunk 4 above)
Comment (shailen14599) [reply src/UKHO.S100PermitService.Common/Handlers/WaitAndRetryPolicy.cs]: Resolved

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Review by harshal11869 — APPROVED** [review]
LGTM.

**Pull request merged**